### PR TITLE
Certificate exporter

### PR DESCRIPTION
**Pull request by gsanchietti**
Export certificates from Traefik.
Traefik module now includes a `certificate-exportert.path` unit. Whenever the `acme.json` file changes, the unit `certificate-exporter.service` is fired. The unit uses `/usr/local/bin/certificate-exporter`, which can be lunched also from command line, like:
```
certificate-exporter /home/traefik1/.local/share/containers/storage/volumes/traefik-acme/_data/acme.json traefik1
```

Exported certificated are saved inside redis as `module/traefik<X>/certificate/<domain>.cert` and  `module/traefik<X>/certificate/<domain>.key`.

TODO:
- [x] inform modules about certificate change, see #53 
- [x] document how to use it

### CODE REVIEW

File: traefik/imageroot/systemd/user/certificate-exporter.service [hunk 1]
```diff
@@ -0,0 +1,12 @@
+[Unit]
+Description=Executes exporter on acme.json changes
+
+[Service]
+Type=simple
+EnvironmentFile=/etc/nethserver/agent.env
+EnvironmentFile=%S/state/environment
+EnvironmentFile=%S/state/agent.env
+ExecStart=/usr/local/bin/certificate-exporter -v %h/.local/share/containers/storage/volumes/traefik-acme/_data/acme.json ${MODULE_ID}
```
Comment (DavidePrincipi) [inline traefik/imageroot/systemd/user/certificate-exporter.service:9]: A system-wide helper script shouldn't be needed. I'd prefer to have a module-local script, e.g.

```suggestion
ExecStart=%S/bin/export-certificate
WorkingDirectory=%h/.local/share/containers/storage/volumes/traefik-acme/_data/
```

The local script implementation can expect to run on the right directory with the right environment.

The above `EnvironmentFile`s should already import the variables used by the script:
- NODE_ID
- MODULE_ID (I'd suggest also AGENT_ID, as it is already prefixed like `module/traefik1`)

File: core/imageroot/usr/local/bin/certificate-exporter [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            cur_key = rdb.hget(rkey, 'key')
+            # save the certificate only if not exists or if has been changed 
+            if (not cur_cert or cur_cert != info["certificate"]) or (not cur_key or cur_key != info["key"]):
+                if verbose:
+                    print(f'Saving certificate and key inside "{rkey}"')
+                rdb.hset(rkey, "cert", info["certificate"])
+                rdb.hset(rkey, "key",  info["key"])
+
+                # signal the certificate-update event
+                event = {"certificate": info["certificate"], "key": info["key"], "node": rdb.hget(f'module/{module_id}/environment', "NODE_ID"), "module": module_id, "domain": info["domain"]}
```
Comment (DavidePrincipi) [inline core/imageroot/usr/local/bin/certificate-exporter:56]: Should be in the environment.

```suggestion
                event = {"certificate": info["certificate"], "key": info["key"], "node": node_id, "module": module_id, "domain": info["domain"]}
```

File: core/imageroot/usr/local/bin/certificate-exporter [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+import os
+import json
+import json
+import agent
+import os.path
+import argparse
+
+def _read_and_set(path, module_id, verbose):
+    base_key = f'module/{args.module_id}/certificate'
```
Comment (DavidePrincipi) [inline core/imageroot/usr/local/bin/certificate-exporter:40]: Fail if environment is not set correctly

```suggestion
    module_id = os.environ['MODULE_ID']
    node_id = os.environ['NODE_ID']
    base_key = f'module/{module_id}/certificate'
```

File: core/imageroot/usr/local/bin/certificate-exporter [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        for info in data["letsencrypt"]["Certificates"]:
+            rkey = f'{base_key}/{info["domain"]["main"]}'
+            cur_cert = rdb.hget(rkey, 'cert')
+            cur_key = rdb.hget(rkey, 'key')
+            # save the certificate only if not exists or if has been changed 
+            if (not cur_cert or cur_cert != info["certificate"]) or (not cur_key or cur_key != info["key"]):
+                if verbose:
+                    print(f'Saving certificate and key inside "{rkey}"')
+                rdb.hset(rkey, "cert", info["certificate"])
+                rdb.hset(rkey, "key",  info["key"])
```
Comment (DavidePrincipi) [inline core/imageroot/usr/local/bin/certificate-exporter:53]: In a single and atomic call:

```suggestion
                rdb.hset(rkey, mapping={"cert": info["certificate"], "key":  info["key"]})
```

File: core/imageroot/usr/local/bin/certificate-exporter [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import json
+import json
+import agent
+import os.path
+import argparse
+
+def _read_and_set(path, module_id, verbose):
+    base_key = f'module/{args.module_id}/certificate'
+    rdb = agent.redis_connect(privileged=True)
+    with open(path) as fp:
```
Comment (DavidePrincipi) [inline core/imageroot/usr/local/bin/certificate-exporter:42]: A non-interactive script could expect to run with the correct `WorkingDirectory=`

```suggestion
    with open('./acme.json') as fp:
```

**Review by DavidePrincipi — COMMENTED** [review]
I think the command should not be invoked manually. I'd rather see it as a module-local and non-interactive script. A manual invocation (if really needed) could be run with

    systemctl --user start certificate-exporter

- no need to document a system-wide command
- the path can be hardcoded
- command line parsing is not needed

**Review by Amygos — APPROVED** [review]

**Review by DavidePrincipi — APPROVED** [review]

**Pull request merged**